src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/FrozenDictionary.cs

```diff
@@ -193,109 +193,115 @@ public static class FrozenDictionary
+                {
+                    return (FrozenDictionary<TKey, TValue>)(object)new Int32FrozenDictionary<TValue>((Dictionary<int, TValue>)(object)source);
                 }
+
+                // Fallback to an implementation usable with any value type and the default comparer.
+                return new ValueTypeDefaultComparerFrozenDictionary<TKey, TValue>(source);
             }
-            else if (typeof(TKey) == typeof(string))
+
+            // Optimize for string keys with the default, Ordinal, or OrdinalIgnoreCase comparers.
+            // If the key is a string and the comparer is known to provide ordinal (case-sensitive or case-insensitive) semantics,
+            // we can use an implementation that's able to examine and optimize based on lengths and/or subsequences within those strings.
+            if (typeof(TKey) == typeof(string) &&
+                (ReferenceEquals(comparer, EqualityComparer<TKey>.Default) || ReferenceEquals(comparer, StringComparer.Ordinal) || ReferenceEquals(comparer, StringComparer.OrdinalIgnoreCase)))
             {
-                // If the key is a string and the comparer is known to provide ordinal (case-sensitive or case-insensitive) semantics,
-                // we can use an implementation that's able to examine and optimize based on lengths and/or subsequences within those strings.
-                if (ReferenceEquals(comparer, EqualityComparer<TKey>.Default) ||
-                    ReferenceEquals(comparer, StringComparer.Ordinal) ||
-                    ReferenceEquals(comparer, StringComparer.OrdinalIgnoreCase))
-                {
-                    Dictionary<string, TValue> stringEntries = (Dictionary<string, TValue>)(object)source;
-                    IEqualityComparer<string> stringComparer = (IEqualityComparer<string>)(object)comparer;
+                Dictionary<string, TValue> stringEntries = (Dictionary<string, TValue>)(object)source;
+                IEqualityComparer<string> stringComparer = (IEqualityComparer<string>)(object)comparer;
 
-                    FrozenDictionary<string, TValue>? frozenDictionary = LengthBucketsFrozenDictionary<TValue>.CreateLengthBucketsFrozenDictionaryIfAppropriate(stringEntries, stringComparer);
-                    if (frozenDictionary is not null)
-                    {
-                        return (FrozenDictionary<TKey, TValue>)(object)frozenDictionary;
-                    }
+                // Calculate the minimum and maximum lengths of the strings in the dictionary. Several of the analyses need this.
+                int minLength = int.MaxValue, maxLength = 0;
+                foreach (KeyValuePair<string, TValue> kvp in stringEntries)
+                {
+                    if (kvp.Key.Length < minLength) minLength = kvp.Key.Length;
+                    if (kvp.Key.Length > maxLength) maxLength = kvp.Key.Length;
+                }
+                Debug.Assert(minLength >= 0 && maxLength >= minLength);
```

Comment:
Will we throw here if provided a `source` of type `Dictionary<string, TValue>` that has ZERO entries? Would we have landed down here eventually or is there some optimization higher up that would have prevented calling down this far? If there is value in such a thing, would we want to short-circuit for any empty source?

Reply:
We won't get here if the collection is empty.  That's handled early on by https://github.com/dotnet/runtime/blob/eea79bb36926b34c62e59d41ad39e8575b0cd613/src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/FrozenDictionary.cs#L135-L142